docs/proposals/application-name-identifier.md

```diff
@@ -0,0 +1,397 @@
+
+### Detailed examples
+
+### Security Considerations
+
+We think this change will not have a direct impact on the security of Argo CD
+or the applications it manages. 
+
+One concern however is that when a user intentionally produces an application
+name whose hash will collide with another, existing application's hash. That
+may lead to the undesired deletion (pruning) of resources that do in fact
+belong to another application outside the adversaries control. However, this
+risk could be properly mitigated by permissions in the `AppProject`, since
+Argo CD will never prune resources that are not allowed for any specific
+Application.
+
+### Risks and Mitigations
+
+The main risks from the authors' points of view are collisions in the hashes
+of application names. For example, if the strings `some-app` and `other-app`
+produce the same hash, this could result in undesired behavior, especially
+when unintentional. The mitigation for this is to chose a hashing algorithm
+with a high resilience against collisions.
+
+Another risk is that third party tools that need to map a managed resource
+in the cluster back to the managing application rely on the clear text value
+of the application instance label. These tools would have to be adapted to
+read the proposed annotation instead of the application instance label.
+
+We think that this may be a breaking change, however, the advantages outweigh
+the risks.
+
+### Upgrade / Downgrade Strategy
+
+Upgrading to a version that implements this proposal should be seamless, as
+previously injected labels will not be removed and additional annotations will
+be applied to the resource. E.g. consider following resource in Git, that will
+be synced as part of an application named `some-application`. In Git, the
+resource looks like follows:
+
+```yaml
+apiVersion: v1
+Kind: Secret
+metadata:
+  name: some-secret
+  namespace: some-namespace
+```
+
+When synced with the current incarnation of Argo CD, Argo CD would inject the
+application instance label and once the resource is applied in the cluster, it
+would look like follows:
+
+```yaml
+apiVersion: v1
+Kind: Secret
+metadata:
+  name: some-secret
+  namespace: some-namespace
+  labels:
+    app.kubernetes.io/instance: some-application
+```
+
+Once Argo CD is updated to a version implementing this proposal, the resource
+would be rewritten to look like the following:
+
+```yaml
+apiVersion: v1
+Kind: Secret
+metadata:
+  name: some-secret
+  namespace: some-namespace
+  labels:
+    app.kubernetes.io/instance: 3fbe5782d494cc956140bc58386c971bf0e96fad
+  annotations:
+    argo-cd.argoproj.io/application-name: some-application
+```
+
+This would be very similar to _renaming_ an application, as in the managed
+resources metadata would be modified in the cluster.
+
+On a rollback to a previous Argo CD version, this change would be reverted
+and the resource would look like the first shown example above.
+
```

Comment:
How would Argo CD know whether a resource's `app.kubernetes.io/instance` was applied using the old system or the new system (the one described in this proposal)?

For example:
1) A cluster is running with the old version of Argo CD
2) All the application resources are labelled using the old system
3) Argo CD is upgraded
4) The new Argo CD looks at all the existing `app.kubernetes.io/instance` fields of the existing resources, discovers that they don't match any expected hash (and thus don't belong to any Application), and ignores them (recreating new Application resources).

I think we need the ability to look at the `app.kubernetes.io/instance` field of a resource and know which system it was created with.

One option we could use is look at the `app.kubernetes.io/instance` value and determine if it is a 32 character hexadecimal value, but this is not foolproof (users may be using Applications with names that are 32-byte hexadecimal values).

Reply:
That's a good point you raise there. 

I haven't yet verified, but I _think_ it would be similar to the behavior if you rename an application, e.g. by deleting an `Application` and re-creating it with a different name, but same parameters (source + destination). 

Argo CD would just go and overwrite what's in `app.kubernetes.io/instance` - at least as long as another application with the previous name does not exist, in which case Argo CD would issue a _Shared Resource Warning_ and prevent the overwrite.

Reply:
I think if an Argo CD instance encounters a resource with a `app.kubernetes.io/instance` value that it doesn't recognize (eg it does not correspond to a hash of a known `application name+install id` value), then Argo CD must assume that that resource is owned by another Argo CD instance (the goal of 'allow multiple Argo CD instances to manage resources on the same cluster').

So, presuming that Argo CD encounters a resource it thinks is owned by someone else, and it has the same name/namespace as a resouce managed by an `Application` managed by the current Argo CD instance, do we:
1) Take ownership of it with the current Argo CD instance
2) Leave it alone

Probably 2), and report an error ('error: Argo CD Application would overwrite a resource that may be owned by another Argo CD instance').

It might be worthwhile to document the algorithm that Argo CD will use to move existing resources from the old style to the new style, in the upgrade scenario.

I'm guessing it would be:
1) If we encounter a resource that contains a `app.kubernetes.io/instance` that is not a 32/64 character hexadecimal value, AND it matches the name field of an existing Argo CD Application, then convert it to the new system (ensure the `app.kubernetes.io/instance` field is updated)
    - As you said, this might already be handled by the rename-style behaviour.
    - Another option would be to delete the resource, and hope that Argo CD recreates it.
2) If we encounter a resource that contains a `app.kubernetes.io/instance` that is a 32/64 character hexadecimal value, leave it alone. (Assume it's a valid instance id from another Argo CD instance).
    - Unless it has the same Kubernetes resource name as a resource that this Argo CD instance is trying to create, in which case report the above error ('error: would overwrite a resource that may be owned...')
3) Else, either delete or leave it alone (orphan it).
(in very rare cases this might orphan or delete old resources, but it's probably fine.)